File: common/src/main/java/rocks/milspecsg/msessentials/common/chat/CommonChatService.java
/*
 *     MSEssentials - MilSpecSG
 *     Copyright (C) 2020 STG_Allen
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package rocks.milspecsg.msessentials.common.chat;

import com.google.inject.Inject;
import com.google.inject.Singleton;
import rocks.milspecsg.msessentials.api.chat.ChatService;
import rocks.milspecsg.msessentials.api.data.config.Channel;
import rocks.milspecsg.msessentials.api.data.key.MSEssentialsKeys;
import rocks.milspecsg.msessentials.api.member.MemberManager;
import rocks.milspecsg.msrepository.api.data.key.Key;
import rocks.milspecsg.msrepository.api.data.registry.Registry;
import rocks.milspecsg.msrepository.api.util.StringResult;
import rocks.milspecsg.msrepository.api.util.UserService;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

@Singleton
public class CommonChatService<
    TPlayer extends TCommandSource,
    TString,
    TCommandSource>
    implements ChatService<TString> {

    @Inject
    Registry registry;

    @Inject
    MemberManager<TString> memberManager;

    @Inject
    StringResult<TString, TCommandSource> stringResult;

    @Inject
    private UserService<TPlayer, TPlayer> userService;

    Map<UUID, String> channelMap = new HashMap<>();
    List<String> playerListMap = new ArrayList<>();

    @Override
    public void switchChannel(UUID userUUID, String channelId) {
        channelMap.put(userUUID, channelId);
    }

    @Override
    public String getChannelIdForUser(UUID userUUID) {
        String channelId = channelMap.get(userUUID);
        if (channelId == null) {
            return registry.getOrDefault(MSEssentialsKeys.CHAT_DEFAULT_CHANNEL);
        }
        return channelId;
    }

    @Override
    public Optional<Channel> getChannelFromId(String channelId) {
        return registry.get(MSEssentialsKeys.CHAT_CHANNELS).flatMap(channels ->
            channels.stream()
                .filter(c -> c.id.equals(channelId))
                .findAny()
        );
    }

    @Override
    public Optional<String> getChannelPrefix(String channelId) {
        return getChannelFromId(channelId).map(c -> c.prefix);
    }

    @Override
    public List<String> getUsersInChannel(String channelId) {
        List<String> userList = new ArrayList<>();
        return userList;
    }

    @Override
    public CompletableFuture<Void> sendMessageToChannel(String channelId, TString message) {
        return CompletableFuture.runAsync(() -> userService.getOnlinePlayers().forEach(p -> {
            if (getChannelIdForUser(userService.getUUID(p)).equals(channelId)) {
                stringResult.send(message, p);
            }
        }));
    }

    @Override
    public CompletableFuture<Void> sendGlobalMessage(TString message) {
        return CompletableFuture.runAsync(() -> userService.getOnlinePlayers().forEach(p -> stringResult.send(message, p)));
    }

    @Override
    public CompletableFuture<TString> formatMessage(
        String prefix,
        String nameColor,
        String userName,
        String message,
        boolean hasChatColorPermission,
        String suffix,
        String serverName,
        String channelId,
        String channelPrefix
    ) {
        return memberManager.getPrimaryComponent().getOneForUser(userName).thenApplyAsync(optionalMember -> {
            if (!optionalMember.isPresent()) {
                return stringResult.fail("Couldn't find a user matching that name!");
            }

            String finalName = optionalMember.get().getUserName();
            if (optionalMember.get().getNickName() != null) {
                finalName = optionalMember.get().getNickName();
            } else {
                finalName = nameColor + finalName;
            }
            return stringResult
                .builder()
                .append(stringResult.deserialize(replacePlaceholders(message, prefix, finalName, hasChatColorPermission, suffix, serverName, channelPrefix, MSEssentialsKeys.PROXY_CHAT_FORMAT_MESSAGE)))
                .onHoverShowText(stringResult.deserialize(replacePlaceholders(message, prefix, finalName, hasChatColorPermission, suffix, serverName, channelPrefix, MSEssentialsKeys.PROXY_CHAT_FORMAT_HOVER)))
                .onClickSuggestCommand(replacePlaceholders(message, prefix, userName, hasChatColorPermission, suffix, finalName, channelPrefix, MSEssentialsKeys.PROXY_CHAT_FORMAT_CLICK_COMMAND))
                .build();
        });
    }

    private String replacePlaceholders(
        String rawMessage,
        String prefix,
        String userName,
        boolean hasChatColorPermission,
        String suffix,
        String serverName,
        String channelPrefix,
        Key<String> key
    ) {
        return registry.get(key)
            .orElseThrow(() -> new IllegalStateException("Missing chat formatting!"))
            .replace("%prefix%", prefix)
            .replace("%username%", userName)
            .replace("%suffix%", suffix)
            .replace("%server%", serverName)
            .replace("%channel%", channelPrefix)
            .replace("%message%", hasChatColorPermission ? rawMessage : stringResult.removeColor(rawMessage));
    }

    @Override
    public String getPlayerList() {
        return userService.getOnlinePlayers().stream().map(userService::getUserName).collect(Collectors.joining(", \n"));
    }

    @Override
    public TString list() {
        return stringResult.builder()
            .green().append("------------------- Online Players --------------------\n")
            .gray().append(getPlayerList())
            .build();
    }
}

File: core/src/main/kotlin/org/anvilpowered/catalyst/core/chat/CommonChatService.kt
/*
 *   Catalyst - AnvilPowered
 *   Copyright (C) 2021
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see https://www.gnu.org/licenses/.
 */

package org.anvilpowered.catalyst.core.chat

import com.google.inject.Inject
import com.google.inject.Singleton
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import org.anvilpowered.anvil.api.Anvil
import org.anvilpowered.anvil.api.misc.sendTo
import org.anvilpowered.anvil.api.misc.sendToConsole
import org.anvilpowered.anvil.api.registry.Registry
import org.anvilpowered.anvil.api.util.PermissionService
import org.anvilpowered.anvil.api.util.SendTextService
import org.anvilpowered.anvil.api.util.UserService
import org.anvilpowered.catalyst.api.ChatMessage
import org.anvilpowered.catalyst.api.config.CatalystKeys
import org.anvilpowered.catalyst.api.registry.CatalystKeys
import org.anvilpowered.catalyst.api.service.ChannelService
import org.anvilpowered.catalyst.api.service.ChatService
import org.anvilpowered.catalyst.api.service.LuckpermsService
import java.util.Locale
import java.util.UUID
import java.util.concurrent.CompletableFuture

@Singleton
class CommonChatService<TPlayer, TCommandSource> @Inject constructor(
    private val channelService: ChannelService<TPlayer>,
    private val luckpermsService: LuckpermsService,
    private val permissionService: PermissionService,
    private val registry: Registry,
    private val userService: UserService<TPlayer, TPlayer>
) : ChatService<TPlayer, TCommandSource> {

    var ignoreMap = mutableMapOf<UUID, MutableList<UUID>>()
    var disabledList = mutableListOf<UUID>()

    override fun sendMessageToChannel(channelId: String, message: Component, senderUUID: UUID): CompletableFuture<Void> {
        return CompletableFuture.runAsync {
            userService.onlinePlayers().forEach {
                if (permissionService.hasPermission(it, registry.get(CatalystKeys.ALL_CHAT_CHANNELS_PERMISSION))
                    || channelService.fromUUID(userService.getUUID(it)).id == channelId
                ) {
                    if (senderUUID != userService.getUUID(it)) {
                        if (!isIgnored(userService.getUUID(it)!!, senderUUID)) {
                            message.sendTo(it)
                        }
                    } else {
                        Anvil.environment?.injector?.getInstance(SendTextService::class.java)?.send(it, message)
                    }
                }
            }
        }
    }

    override fun ignore(playerUUID: UUID, targetPlayerUUID: UUID): Component {
        val targetPlayer = userService.getPlayer(targetPlayerUUID) ?: return Component.text("That user does not exist!").color(NamedTextColor.RED)
        var uuidList: MutableList<UUID> = ArrayList()
        if (ignoreMap[playerUUID] == null) {
            uuidList.add(targetPlayerUUID)
        } else {
            uuidList = ignoreMap[playerUUID] ?: return Component.text("An error occurred.").color(NamedTextColor.RED)
            if (uuidList.contains(targetPlayerUUID)) {
                return unIgnore(playerUUID, targetPlayerUUID)
            }
        }
        ignoreMap[playerUUID] = uuidList
        return Component.text()
            .append(Component.text("You are now ignoring ").color(NamedTextColor.GREEN))
            .append(Component.text(userService.getUserName(targetPlayer)).color(NamedTextColor.GOLD))
            .build()
    }

    override fun unIgnore(playerUUID: UUID, targetPlayerUUID: UUID): Component {
        val uuidList: MutableList<UUID> = ignoreMap[playerUUID] ?: return Component.text("That user does not exist!").color(NamedTextColor.RED)
        if (isIgnored(playerUUID, targetPlayerUUID)) {
            uuidList.remove(targetPlayerUUID)
            ignoreMap.replace(playerUUID, uuidList)
        }
        return Component.text("You are no longer ignoring ${userService.getUserName(targetPlayerUUID)}").color(NamedTextColor.GREEN)
    }

    override fun isIgnored(playerUUID: UUID, targetPlayerUUID: UUID): Boolean {
        val uuidList = ignoreMap[playerUUID] ?: return false
        return targetPlayerUUID in uuidList
    }

    override fun checkPlayerName(sender: TPlayer, msg: String): String {
        var message = msg
        for (player in userService.onlinePlayers()) {
            val username = userService.getUserName(player)
            if (message.lowercase(Locale.getDefault()).contains(username.lowercase(Locale.getDefault()))) {
                val occurrences: MutableList<Int> = ArrayList()
                var startIndex = message.lowercase(Locale.getDefault()).indexOf(username.lowercase(Locale.getDefault()))
                while (startIndex != -1) {
                    occurrences.add(startIndex)
                    startIndex = message.lowercase(Locale.getDefault()).indexOf(username.lowercase(Locale.getDefault()), startIndex + 1)
                }
                val chatColor = luckpermsService.getChatColor(sender)
                for (occurrence in occurrences) {
                    message = message.substring(0, occurrence) + "&b@" + username + chatColor + message.substring(occurrence + username.length)
                }
            }
        }
        return message
    }

    override fun sendChatMessage(chatMessage: ChatMessage) {
        if (chatMessage.component == Component.text("")) {
            return
        }
        chatMessage.component.sendToConsole<TCommandSource>()
        sendMessageToChannel(channelService.fromUUID(chatMessage.uuid).id, chatMessage.component, chatMessage.uuid)
    }

    override fun toggleChatForUser(player: TPlayer) {
        val userUUID = userService.getUUID(player)
        if (!disabledList.contains(userUUID)) {
            disabledList.add(userUUID!!)
            return
        }
        disabledList.remove(userUUID)
    }

    override fun isDisabledForUser(player: TPlayer): Boolean {
        val userUUID = userService.getUUID(player)
        return disabledList.contains(userUUID)
    }
}
